### PR TITLE
Add a MolBundle class

### PR DESCRIPTION
**Pull request by greglandrum**
#### What does this implement/fix? Explain your changes.
This adds a `MolBundle`: a class to hold groups of molecules that have the topology (currently enforced as the same number of atoms and bonds).
It also adds integration of that class with the substructure searching code.

At the moment this isn't particular interesting, but it lays the groundwork for support of things like enhanced stereochemistry in the not-too-distant future.

#### Any other comments?
The current implementation is certainly not the most memory efficient way of solving the problem, but the API should allow us to move to something better if that becomes important. The once concern there is the `getMols()` method, which we may want to consider removing from the API.

### CODE REVIEW

**Comment by bp-kelley** [comment]
@greglandrum Did my comments appear before I requested changes?  These might have been sitting for a while...

**Comment by greglandrum** [comment]
no, I'm pretty sure comments don't appear until after the review is submitted

File: Code/GraphMol/MolBundle.h [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  MolBundle() : RDProps(){};
+
+  //! copy constructor
+  MolBundle(const MolBundle &other) : RDProps(other) { d_mols = other.d_mols; };
+  // FIX: need serialization/deserialization
+
+  virtual ~MolBundle(){};
+
+  //! returns our molecules
+  virtual const std::vector<boost::shared_ptr<ROMol> > &getMols() const {
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/MolBundle.h]: I was careless with `virtual` here. After thinking about it, most (all?) methods should probably be virtual: we don't want to constrain how molecules are stored in a bundle in general.

File: Code/GraphMol/MolBundle.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      // REVIEW: should we allow different numbers of bonds?
+      if (nmol->getNumBonds() != d_mols[0]->getNumBonds())
+        throw ValueErrorException(
+            "all molecules in a bundle must have the same number of atoms");
+    }
+    d_mols.push_back(nmol);
+    return (d_mols.size());
+  }
+  //! returns the number of molecules from the bundle
+  unsigned int size() const { return d_mols.size(); };
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/MolBundle.h]: switching to return `size_t` instead. That's more explicit

File: Code/GraphMol/MolBundle.h [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  //! returns a particular molecule in the bundle
+  const boost::shared_ptr<ROMol> getMol(size_t idx) const {
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/MolBundle.h]: fixed. :-)

File: Code/GraphMol/Substruct/SubstructMatch.h [hunk 4]
```diff
@@ -126,6 +127,50 @@ unsigned int SubstructMatch(ResonanceMolSupplier &resMolSuppl,
...
+                    bool useChirality = false,
+                    bool useQueryQueryMatches = false);
+//! \overload
+//! finds the first match in the bundle
+bool SubstructMatch(const MolBundle &bundle, const MolBundle &query,
+                    MatchVectType &matchVect, bool recursionPossible = true,
+                    bool useChirality = false,
+                    bool useQueryQueryMatches = false);
+//! \overload
+//! finds all matches in the first element of the bundle that has any matches
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/Substruct/SubstructMatch.h]: yeah, describing this is non-trivial. will try different wording

**Pull request merged**